warehouse/scripts/publish.py

```diff
@@ -0,0 +1,398 @@
+
+                    fpath = os.path.join(tmpdir, destination.filename(model_name))
+
+                    df = pd.read_gbq(
+                        str(node.select(engine)),
+                        project_id=project,
+                        progress_bar_type="tqdm",
+                    )
+                    df.to_csv(fpath, index=False)
+                    typer.secho(
+                        f"selected {len(df)} rows ({humanize.naturalsize(os.stat(fpath).st_size)}) from {node.schema_table}"
+                    )
+
+                    hive_path = destination.hive_path(exposure, model_name, bucket)
+
+                    msg = f"writing {model_name} to {hive_path} and {destination.url} {ckan_id}"
+                    if dry_run:
+                        typer.secho(
+                            f"would be {msg}",
+                            fg=typer.colors.MAGENTA,
+                        )
+                    else:
+                        typer.secho(msg, fg=typer.colors.GREEN)
+                        fs = gcsfs.GCSFileSystem(
+                            project=project, token="google_default"
+                        )
+                        fs.put(fpath, hive_path)
+
+                        with open(fpath, "rb") as fp:
+                            requests.post(
+                                destination.url,
+                                data={"id": ckan_id},
+                                headers={"Authorization": API_KEY},
+                                files={"upload": fp},
+                            ).raise_for_status()
+            elif isinstance(destination, TileServerDestination):
+                # the depends_on each create a layer
+                for model in exposure.depends_on.nodes:
+                    node = BaseNode._instances[model]
+
+                    geojson_fpath = os.path.join(tmpdir, f"{node.name}.geojson")
+                    fpath = os.path.join(tmpdir, destination.filename(node.name))
+
+                    df = pd.read_gbq(
+                        str(node.select(engine)),
+                        project_id=project,
+                        progress_bar_type="tqdm",
+                    )
+                    # typer.secho(
+                    #     f"selected {len(df)} rows ({humanize.naturalsize(os.stat(fpath).st_size)}) from {model_name}"
+                    # )
+
+                    def make_linestring(x):
+
+                        # shapely errors if the array contains only one point
+                        if len(x) > 1:
+                            # each point in the array is wkt
+                            # so convert them to shapely points via list comprehension
+                            as_wkt = [shapely.wkt.loads(i) for i in x]
+                            return shapely.geometry.LineString(as_wkt)
+
+                    # apply the function
+                    df["geometry"] = df.pt_array.swifter.apply(make_linestring)
```

Comment:
this stuff seems hard coded based on the dim_shapes_geo table specifically... is that right/appropriate? and also is `make_linestring` copied from somewhere? just wondering if we want to include like a reference on that for context or anything

Reply:
Adding a comment. I need to circle back to this (i.e. map tiles) after our HQTA discussion.